darts/utils/data/shifted_dataset.py

```diff
@@ -459,59 +463,123 @@ def __getitem__(self, idx) -> Tuple[np.ndarray, Optional[np.ndarray], np.ndarray
+        time series and the (start, end) indices of all sub sets.
+        This allows to calculate the sub set indices for all future samples `i` by simply adjusting for the difference
+        between the positions of sample `i` and sample `0`.
+
+        Parameters
+        ----------
+        ts_idx
+            index of the current target TimeSeries.
+        ts_target
+            current target TimeSeries.
+        end_of_output_idx
+            the index where the output chunk of the current sample ends in `ts_target`.
+        ts_target
+            current target TimeSeries.
+        ts_covariate
+            current covariate TimeSeries.
+        """
+
+        cov_start, cov_end = None, None
+
+        # the first time ts_idx is observed
+        if ts_idx not in self.index_memory:
+            start_of_output_idx = end_of_output_idx - self.output_chunk_length
+            start_of_input_idx = start_of_output_idx - self.shift
+
+            # select forecast point and target period, using the previously computed indexes
+            future_start, future_end = start_of_output_idx, start_of_output_idx + self.output_chunk_length
+
+            # select input period; look at the `input_chunk_length` points after start of input
+            past_start, past_end = start_of_input_idx, start_of_input_idx + self.input_chunk_length
+
+            if self.covariates is not None:
+                start = future_start if self.shift_covariates else past_start
+                end = future_end if self.shift_covariates else past_end
+
+                # we need to be careful with getting ranges and indexes:
+                # to get entire range, full_range = ts[:len(ts)]; to get last index: last_idx = ts[len(ts) - 1]
+
+                # extract actual index value (respects datetime- and integer-based indexes; also from non-zero start)
+                start_time = ts_target.time_index[start]
+                end_time = ts_target.time_index[end - 1]
+
+                raise_if_not(start_time in ts_covariate.time_index and end_time in ts_covariate.time_index,
+                             f'Missing covariates; could not find {"future" if self.shift_covariates else "past"} '
+                             f'covariates in index value range: {start_time} - {end_time}.')
+
+                # extract the index position (index) from index value
+                cov_start = ts_covariate.time_index.get_loc(start_time)
+                cov_end = ts_covariate.time_index.get_loc(end_time) + 1
+
+            # store position of initial sample and all relevant sub set indices
+            self.index_memory[ts_idx] = {
+                'end_of_output_idx': end_of_output_idx,
+                'past_target': (past_start, past_end),
+                'future_target': (future_start, future_end),
+                'covariate': (cov_start, cov_end),
+            }
+        else:
```

Comment:
Do you know (roughly) what is the fraction of the gain coming from this caching?
Is this avoiding some expensive O(N) `get_loc()` calls?

Reply:
Yes, I did it experimentally, see below.

- n_calls is the number of times we go through all indexes of `__getitem()__`  from a GenericShiftedDataset (n_calls = 3 would be the equivalent of going through 1000 epochs)
- `time` is the time spent for computing `memory_indexer()`

![cach_vs_no_cache](https://user-images.githubusercontent.com/84317528/142828373-a086576e-ead0-48e6-864f-bb3eb6af2db0.png)